chemprop/uncertainty/calibrator.py

```diff
@@ -76,38 +76,148 @@ def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> Sel
 
 @UncertaintyCalibratorRegistry.register("zscaling")
 class ZScalingCalibrator(RegressionCalibrator):
+    """Calibrate regression datasets by applying a scaling value to the uncalibrated standard deviation,
+    fitted by minimizing the negative-log-likelihood of a normal distribution around each prediction. [levi2022]_
+
+    References
+    ----------
+    .. [levi2022] Levi, D.; Gispan, L.; Giladi, N.; Fetaya, E. "Evaluating and Calibrating Uncertainty Prediction in Regression Tasks." Sensors, 2022, 22(15), 5540. https://www.mdpi.com/1424-8220/22/15/5540.
+    """
+
     def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> Self:
-        return self
+        scalings = []
+        for j in range(uncs.shape[1]):
+            mask_j = mask[:, j]
+            preds_j = preds[:, j][mask_j].numpy()
+            uncs_j = uncs[:, j][mask_j].numpy()
+            targets_j = targets[:, j][mask_j].numpy()
+            errors = preds_j - targets_j
 
-    def apply(self, uncs: Tensor) -> Tensor:
-        return
+            def objective(scaler_value: float):
+                scaled_vars = uncs_j * scaler_value**2
+                nll = np.log(2 * np.pi * scaled_vars) / 2 + errors**2 / (2 * scaled_vars)
+                return nll.sum()
 
+            zscore = errors / np.sqrt(uncs_j)
+            initial_guess = np.std(zscore)
+            scalings.append(fmin(objective, x0=initial_guess, disp=False))
 
-@UncertaintyCalibratorRegistry.register("tscaling")
-class TScalingCalibrator(RegressionCalibrator):
-    def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> Self:
+        self.scalings = torch.tensor(scalings)
         return self
 
     def apply(self, uncs: Tensor) -> Tensor:
-        return
+        return uncs * self.scalings**2
 
 
 @UncertaintyCalibratorRegistry.register("zelikman-interval")
 class ZelikmanCalibrator(RegressionCalibrator):
+    """Calibrate regression datasets that does not depend on a particular probability function form.
+
+    This method is designed to be used with interval output. Uses the "CRUDE" method as described in [zelikman2020]_.
```

Comment:
can you add that this currently only works for variance?

Reply:
added